### ISSUE TITLE
New rule: `require-event-dispatcher-types`

### ISSUE THREAD

**Issue by marekdedic** [ota-meshi/eslint-plugin-svelte#346]
### Motivation

Hi,
when using svelte with TypeScript, it is possible to add types to events by typing the `createEventDispatcher` function so:

```svelte
import { createEventDispatcher } from "svelte";

const dispatch = createEventDispatcher<{one: never, two: number}>();

dispatch("one"); // OK
dispatch("two", 3); // OK
dispatch("one", "Hello world"); // Fails
dispatch("two"); // Fails
dispatch("two", "Hello world"); // Fails
```

I propose a rule that would require all calls to `createEventDispatcher` to include the templated type.

### Description

The rule would trigger on any call of `createEventDispatcher` witchou the templated type

### Examples

```svelte
import { createEventDispatcher } from "svelte";

<!-- ✓ GOOD -->
const dispatch = createEventDispatcher<{one: never, two: number}>();

const dispatch = createEventDispatcher<any>();

const dispatch = createEventDispatcher<{}>();

<!-- ✗ BAD -->
const dispatch = createEventDispatcher();
```

**Comment by marekdedic** [comment]
Oh and if you type your events, all the listeners get typechecked as well, so this really is something you want to do if you are using TS...

**Comment by ota-meshi** [comment]
Thanks for the rule suggestions.
That rule sounds almost good to me.

But why is the example below marked as `GOOD`? I think they are pretty much the same if you omit the type.

```js
const dispatch = createEventDispatcher<any>();

const dispatch = createEventDispatcher<{}>();
```

**Comment by marekdedic** [comment]
Hmm, originally, I though of the rule as being a reminder not to forget the types - that's why I put those examples there, I didn't really care about what the type is, just that there is one...

I think `any` doesn't really make sense - it is the same as if you omit the type - so putting `any` in there or manually suppressing the rule does the same thing.

As with the `{}` type - I meant `Record<string, never>`, an empty object (this part of TS is a bit confusing...). But that would make it so that no events can be dispatched, so the whole call is useless and doesn't make any sense...

The rule could report for both of those - I think it probably should report for `any`, with the empty object I wouldn't bother - that's a silly example not worth implementing a check for...

**Comment by ota-meshi** [comment]
Thank you for your opinion.

> I though of the rule as being a reminder not to forget the types I though of the rule as being a reminder not to forget the types

I hear you and I agree with it. The `any` and `{}` may be intentional for the user.
I think it makes sense not to check them to avoid complicating the implementation of the rule.
If there is user feedback that they want to check strictly, we can add the option later.

**Comment by marekdedic** [comment]
Great, yeah, let's make it simple if there's no need to make the rule more complex...
